drv/fpga-server/src/main.rs

```diff
@@ -0,0 +1,272 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Server for managing FPGA devices.
+
+#![no_std]
+#![no_main]
+
+use ringbuf::*;
+use userlib::*;
+use zerocopy::{byteorder, AsBytes, Unaligned, U16};
+
+use drv_fpga_api::*;
+use drv_fpga_devices::{ecp5, ecp5::Ecp5, ecp5_spi::Ecp5UsingSpi, Fpga};
+use drv_spi_api::{self as spi_api, Spi, SpiDevice};
+use drv_stm32xx_sys_api::{self as sys_api, Sys};
+use idol_runtime::{ClientError, Leased, LenLimit, R, W};
+
+task_slot!(SYS, sys);
+task_slot!(SPI, spi_driver);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Trace {
+    None,
+    DeviceId(u32),
+    StartBitstreamLoad(BitstreamType),
+    ContinueBitstreamLoad(usize),
+    FinishBitstreamLoad,
+}
+ringbuf!(Trace, 16, Trace::None);
+
+#[export_name = "main"]
+fn main() -> ! {
+    let sys = Sys::from(SYS.get_task_id());
+    let spi = Spi::from(SPI.get_task_id()).device(0);
+
+    cfg_if::cfg_if! {
+        if #[cfg(target_board = "sidecar-1")] {
+            let driver = Ecp5UsingSpi {
+                sys,
+                spi,
+                done: sys_api::Port::J.pin(15),
+                init_n: sys_api::Port::J.pin(12),
+                program_n: sys_api::Port::J.pin(13),
+                design_reset_n: sys_api::Port::J.pin(14),
+            };
+            driver.configure_gpio();
+            let ecp5 = Ecp5::from(driver);
+        } else if #[cfg(target_board = "gimletlet-2")] {
+            let driver = Ecp5UsingSpi {
+                sys,
+                spi,
+                done: sys_api::Port::E.pin(15),
+                init_n: sys_api::Port::D.pin(12),
+                program_n: sys_api::Port::B.pin(10),
+                design_reset_n: sys_api::Port::D.pin(11),
+            };
+            driver.configure_gpio();
+            let ecp5 = Ecp5::from(driver);
+        } else {
+            compile_error!("Board is not supported by the task/fpga");
+        }
+    }
+
+    let mut incoming = [0u8; idl::INCOMING_SIZE];
+    let mut server = ServerImpl {
+        device: ecp5,
+        device_reset_ticks: ecp5::DEVICE_RESET_DURATION,
+        application: Spi::from(SPI.get_task_id()).device(1),
```

Comment:
`application` is a confusing name for the SPI peripheral, I'd normally pick something like `spi`.

Reply:
This has been renamed to `user_design` throughout the diff.